### DISCUSSION TITLE
Split detekt in different tasks

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#6085] [Ideas]
Right now detekt is a single unit of execution. I think that for 2.0 we should change 2 mayor things.

**First:** We shouldn't execute the `cli` from the Gradle Plugin. Those two should be two different clients of the same code. They shouldn't depend one on each other. Right now this creates us the problems that to support something on the Gradle Plugin we first need to implement it on the `cli`.

**Second:** Detekt should provide a set of tools and each client should compose them to archieve the desired behaviour. My idea is to have a function that checks if the configuration is correct or not, other to analyze the code and create a report and another(s) to generate the desired output for the user (given a report). The idea is that the clients (`cli`, Gradle Plugin, Maven, bazel...) should implement the coordination of these functions to make it work as expected.

The reasons I have to propouse this change:
- A lot of flexibility to implement the clients
  - On the `cli` all of those task will be executed sequentially
  - On the Gradle Plugin we will generate a different task for each individual task and make them depend on each other so we can relay on the incremental and chache features of Gradle.
    - Did you run the detekt task and want to generate the baseline for those? It will go really fast!
    - Did the CI found some detekt issues but you want to see the `html` report on local? No problem! Really fast too because you can download the build-cache for the slow task of detekt and just generate the html report on local.
    - Did you lose the report from the console? No problem, execute the task again, the reports to console will be never `UP-TO-DATE`.
  - Our compiler plugin will just analyze the code. The other parts would be moved outside the compilation phase. This will help with #6084

What I think we need to work on to make this happen:
- Define the analyzer report. Our current `class Findings` containts references to PSI elements. That's difficult to serialize.
- Define the steps/tasks that we want to have. I clearly see check config, analyze and reports. But may be more.
  - For example we could have a serialize/deserialize function. This way the ananlyzer could return a proper instance, the `cli` would pass it directly to the reporters but the Gradle Plugin will call the serializer and on the other tasks call the deserializer.

I would like to hear your feedback :)

**Comment by marschwar** [comment]
I am not sure I understand enough about the inner workings of gradle but it sounds quite nice. I really like to idea to run detekt and store the findings as task output and generate the baseline and reports based on that output.
I only worry about yet another explosion of tasks. For type resolution on android and multi platform there already are more than enough.

> Our compiler plugin will just analyze the code. The other parts would be moved outside the compilation phase.

Does that mean we would have a task `:detektReportHtml` (or something similar) which would depend on `:compileKotlin` and the compiler plugin writes its findings to a well known location and the report generator picks it up there?

**Reply by BraisGabin** [reply]
> Does that mean we would have a task :detektReportHtml (or something similar) which would depend on :compileKotlin and the compiler plugin writes its findings to a well known location and the report generator picks it up there?

Yes, exactly that.

**Comment by 3flex** [comment]
> Right now this creates us the problems that to support something on the Gradle Plugin we first need to implement it on the cli

Is this actually a problem though? CLI should theoretically support everything the Gradle plugin does. I do agree this should change though, probably Gradle plugin should integrate with detekt-tooling instead - I understand the intention is for that to be the primary integration point for all clients, and is used by CLI and IntelliJ plugin already. This could be added to the detekt 2.0 list as its own item.

RE: config check, analysis, report - I don't agree that config check needs its own task, that should be a very quick operation, so there won't be much value splitting that out. The analysis & reporting can be split though and like you say we need to serialise the detekt findings as an analysis output, for this, as well as report merging.

However I agree with the above comment. This will lead to an explosion of tasks and confusion for users. If we have a TXT, HTML, XML, MD, SARIF & console report task generated for every single analysis task it's going to get completely out of hand and confusing for users. We need serialisation for report merging but I don't know about the usability of this approach.

**Reply by BraisGabin** [reply]
About the explosion of tasks: yes, we will have more tasks. But a lot of them would be internal. No one should be interested in execute them. The "public" tasks should be the same. The idea is that you define which reports you want and our task detekt will depend on the necessary tasks to make the job.

> I don't agree that config check needs its own task, that should be a very quick operation, so there won't be much value splitting that out.

The config of detekt changes very little and it is checked for every module twice (main and test). In a project with 100 subproject we will go from 200 checks to 1. And then that 1 will hit UP-TO-DATE every time. But I understand your point. Maybe the configuration of a gradle task takes longer than check it 200 times. If that's the case the gradle plugin will execute it in the same task as the analysis.
